File: apps/frontend/src/lib/encValidation.ts
import type { Map as MapLibreMap, MapGeoJSONFeature, PointLike } from "maplibre-gl";
import type { HazardSummary } from "@revolt/shared-types";
import type { LegPositions } from "./geo.js";

// "lndare" (land area) was missing here from the start; the check only inferred hazard from
// depare/resare/obstrn/uwtroc, so a waypoint on land was never detected. Confirmed by testing an
// unambiguous inland point across zoom 8-14: the pre-fix check reported "safe" every time.
const HAZARD_LAYERS = ["depare", "resare", "obstrn", "uwtroc", "lndare"];

// Chart coverage extent (S-57 M_COVR, CATCOV=1 available / CATCOV=2 no coverage), queried
// separately from the hazard layers above: "no hazard found nearby" and "no chart data exists
// here at all" are very different facts, conflating them would report unsurveyed water as
// confidently "safe". Never painted in the MapLibre style, just present in the tileset so
// queryRenderedFeatures() can see it (a zero-opacity fill layer is still queryable).
const COVERAGE_LAYER = "m_covr";

// Web Mercator meters-per-pixel at zoom 0, latitude 0 (Earth's circumference / 256px tile width);
// the standard constant for converting a real-world distance into an on-screen pixel size that
// scales correctly with zoom (pixel density doubles every zoom level, and shrinks with cos(lat)).
const METERS_PER_PIXEL_AT_ZOOM_0 = 156543.03392;

// Real-world half-width (m) of the safety corridor checked/shown around each leg: a small, fixed
// margin appropriate for a vessel this size, not a mariner-configurable XTD/beam setting (see
// SAFETY_CONTOUR_M's sibling comment in MapWidget.tsx). Exported so MapWidget's corridor
// visualization draws exactly the real-world band this check covers.
export const CORRIDOR_HALF_WIDTH_M = 15;

/**
 * Converts a real-world distance (m) to on-screen pixels at a given zoom/latitude. A fixed pixel
 * padding meant something different at every zoom level; this recomputes it from a real-world
 * distance instead, matching how a real ECDIS XTD corridor behaves.
 */
export function metersToPixels(meters: number, zoom: number, latitude: number): number {
	const metersPerPixel =
		(METERS_PER_PIXEL_AT_ZOOM_0 * Math.cos((latitude * Math.PI) / 180)) / Math.pow(2, zoom);
	return meters / metersPerPixel;
}

interface DepareProperties {
	DRVAL1?: number;
}

interface CoverageProperties {
	CATCOV?: number;
}

function isCoverageFeature(feature: MapGeoJSONFeature): boolean {
	return (
		feature.layer.id === COVERAGE_LAYER &&
		(feature.properties as CoverageProperties | null)?.CATCOV === 1
	);
}

// Checked at each endpoint with its own small box, not the shared bbox spanning both endpoints
// used for the hazard layers below: a box spanning both can overlap the covered endpoint's own
// polygon even when the far endpoint is outside coverage, silently reporting the whole leg
// covered whenever just one end was.
function isCoveredAt(
	map: MapLibreMap,
	point: { x: number; y: number },
	paddingPx: number,
): boolean {
	// If the layer doesn't currently exist in the style (mid theme-transition, see the hazard-layer
	// comment below), skip the check rather than concluding "not covered": a missing layer means
	// "this evaluation finds nothing new", not a false negative.
	if (!map.getLayer(COVERAGE_LAYER)) return true;
	const box: [PointLike, PointLike] = [
		[point.x - paddingPx, point.y - paddingPx],
		[point.x + paddingPx, point.y + paddingPx],
	];
	const features = map.queryRenderedFeatures(box, { layers: [COVERAGE_LAYER] });
	return features.some(isCoverageFeature);
}

function isDepareHazard(feature: MapGeoJSONFeature, safetyContourM: number): string | null {
	const drval1 = (feature.properties as DepareProperties | null)?.DRVAL1;
	if (typeof drval1 === "number" && drval1 < safetyContourM) {
		return `Leg crosses charted depth below the ${String(safetyContourM)} m safety contour (DRVAL1=${String(drval1)} m).`;
	}
	return null;
}

const RESTRICTED_LAYER_DESCRIPTIONS: Record<string, string> = {
	resare: "Leg passes through a charted restricted area.",
	obstrn: "Leg passes near a charted obstruction.",
	uwtroc: "Leg passes near a charted underwater rock.",
	lndare: "Leg crosses charted land.",
};

function evaluateLeg(map: MapLibreMap, leg: LegPositions, safetyContourM: number): HazardSummary {
	const p1 = map.project([leg.from.lon, leg.from.lat]);
	const p2 = map.project([leg.to.lon, leg.to.lat]);
	const avgLat = (leg.from.lat + leg.to.lat) / 2;
	const paddingPx = metersToPixels(CORRIDOR_HALF_WIDTH_M, map.getZoom(), avgLat);
	const bbox: [PointLike, PointLike] = [
		[Math.min(p1.x, p2.x) - paddingPx, Math.min(p1.y, p2.y) - paddingPx],
		[Math.max(p1.x, p2.x) + paddingPx, Math.max(p1.y, p2.y) + paddingPx],
	];

	// queryRenderedFeatures throws (not just "no results") if a named layer doesn't currently exist
	// in the style, which briefly happens mid-transition (setStyle()'s day/dusk swap synchronously
	// clears every layer until "style.load" reloads it). Filtering to layers that actually exist
	// degrades gracefully instead of throwing if a recompute lands in that window.
	const availableHazardLayers = HAZARD_LAYERS.filter((id) => map.getLayer(id));
	const features = map.queryRenderedFeatures(bbox, { layers: availableHazardLayers });

	for (const feature of features) {
		const description = RESTRICTED_LAYER_DESCRIPTIONS[feature.layer.id];
		if (description) {
			return { status: "blocked", description };
		}
	}
	// A bounding-box existence check per endpoint, not the true containment check Phase 2's
	// ST_Covers does server-side; consistent with every other check here being an approximation
	// of what's currently rendered, not authoritative. Still approximate even per-endpoint: a
	// coverage gap in the middle of a long leg with both endpoints covered would slip through;
	// Phase 2 catches that at send time regardless.
	if (!isCoveredAt(map, p1, paddingPx) || !isCoveredAt(map, p2, paddingPx)) {
		return {
			status: "no_data",
			description: "No charted ENC data covers this leg; not verified safe, just unchecked.",
		};
	}
	for (const feature of features) {
		if (feature.layer.id === "depare") {
			const warning = isDepareHazard(feature, safetyContourM);
			if (warning) return { status: "warning", description: warning };
		}
	}
	return { status: "safe", description: "No charted hazards near this leg." };
}

/**
 * Client-side ENC hazard check (Phase 1) against the already-rendered depare/resare/obstrn/uwtroc/
 * lndare vector tile layers, plus m_covr for chart coverage extent. Advisory only: only sees
 * what's currently rendered at the current zoom, with tile-simplified geometry. The authoritative
 * server-side PostGIS check is Phase 2.
 */
export function evaluateEncHazards(
	map: MapLibreMap,
	legs: LegPositions[],
	safetyContourM: number,
): Record<string, HazardSummary> {
	const result: Record<string, HazardSummary> = {};
	for (const leg of legs) {
		result[leg.toId] = evaluateLeg(map, leg, safetyContourM);
	}
	return result;
}

File: apps/backend/src/revolt_api/enc_validation.py
"""Phase 2 (server-side, authoritative) ENC hazard validation.

Companion to apps/frontend/src/lib/encValidation.ts's Phase 1 client-side check: same five hazard
layers (HAZARD_LAYERS there, _BLOCKED_LAYERS + "depare" here), same safety_contour_m depth
threshold, but queries the enc_* PostGIS tables (infra/enc-pipeline/ingest_postgis.sh) directly
instead of whatever happens to be rendered in the browser's current viewport/zoom -- see
WebApp/CLAUDE.md's ENC validation section for why that distinction matters.

The route geometry checked is the vessel's actual path (straight legs joined by turn-radius arcs,
geo.py's build_route_points), not a naive straight-line-through-every-waypoint polyline, so this
agrees with what MapWidget draws rather than being a cruder approximation of it.

Also checks chart coverage (enc_m_covr, S-57's CATCOV=1/2 metadata layer) separately from the
hazard layers -- "no charted hazard found nearby" and "no chart data exists here at all" are very
different facts, and conflating them would let a route through completely unsurveyed water report
as confidently "safe". Unlike "blocked", "no_data" does NOT refuse a send: the vessel is tested in
areas outside this delivery's ENC coverage, so sending has to stay possible there.
"""

from collections.abc import Sequence
from typing import Literal

from geoalchemy2.shape import to_shape
from pydantic import BaseModel
from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession

from revolt_api.geo import RouteWaypoint, build_route_points
from revolt_api.models.mission import Waypoint

ValidationStatus = Literal["safe", "warning", "no_data", "blocked"]

# Layers that always block a route outright if the buffered path comes within safety_margin_m --
# matches encValidation.ts's RESTRICTED_LAYER_DESCRIPTIONS exactly (must never disagree with
# Phase 1 about what counts as a hazard).
_BLOCKED_LAYER_DESCRIPTIONS: dict[str, str] = {
	"resare": "Route passes through a charted restricted area.",
	"obstrn": "Route passes near a charted obstruction.",
	"uwtroc": "Route passes near a charted underwater rock.",
	"lndare": "Route crosses charted land.",
}
_DEPTH_LAYER = "depare"
_NO_DATA_DESCRIPTION = "Route passes through an area with no charted ENC data."


class HazardHit(BaseModel):
	layer: str
	description: str
	count: int


class ValidationResult(BaseModel):
	status: ValidationStatus
	hazards: list[HazardHit]


def _route_linestring_wkt(waypoints: Sequence[Waypoint]) -> str | None:
	route_waypoints: list[RouteWaypoint] = []
	for wp in waypoints:
		point = to_shape(wp.position)
		route_waypoints.append(RouteWaypoint(lat=point.y, lon=point.x, switch_radius=wp.switch_radius))
	points = build_route_points(route_waypoints)
	if len(points) < 2:
		return None
	# WKT is (x y) = (lon lat), the opposite order from how we store/pass lat/lon everywhere else.
	coords = ", ".join(f"{lon} {lat}" for lat, lon in points)
	return f"LINESTRING({coords})"


async def _count_nearby(
	db: AsyncSession,
	table: str,
	route_wkt: str,
	margin_m: float,
	max_depth_m: float | None = None,
) -> int:
	# table is one of the fixed literals in _BLOCKED_LAYER_DESCRIPTIONS/_DEPTH_LAYER above, never
	# user input, so this f-string can't be used for SQL injection despite not being a bind param.
	# Everything that IS a runtime value (margin_m, max_depth_m, route_wkt) goes through a bind
	# parameter regardless of how it's sourced.
	depth_clause = "drval1 < :max_depth_m AND " if max_depth_m is not None else ""
	params: dict[str, object] = {"route_wkt": route_wkt, "margin_m": margin_m}
	if max_depth_m is not None:
		params["max_depth_m"] = max_depth_m
	result = await db.execute(
		text(
			f"SELECT count(*) FROM {table} "
			f"WHERE {depth_clause}"
			"ST_DWithin(geom::geography, ST_GeomFromText(:route_wkt, 4326)::geography, :margin_m)"
		),
		params,
	)
	return result.scalar_one()


async def _route_lacks_coverage(db: AsyncSession, route_wkt: str, margin_m: float) -> bool:
	"""True if any part of the buffered route falls outside the union of CATCOV=1 (coverage
	available) polygons -- a real containment check (ST_Covers), not an existence/nearby check
	like _count_nearby's, since "some coverage polygon happens to be nearby" is a different and
	weaker claim than "the whole route is actually within surveyed area"."""
	result = await db.execute(
		text(
			"SELECT NOT ST_Covers("
			"  (SELECT ST_Union(geom) FROM enc_m_covr WHERE catcov = 1),"
			"  ST_Buffer(ST_GeomFromText(:route_wkt, 4326)::geography, :margin_m)::geometry"
			")"
		),
		{"route_wkt": route_wkt, "margin_m": margin_m},
	)
	lacks_coverage = result.scalar_one()
	# NOT ST_Covers(NULL, ...) is SQL NULL, not a boolean, when enc_m_covr has zero CATCOV=1 rows
	# (e.g. ingestion never run) -- treat "we have no coverage information at all" as lacking
	# coverage too, rather than silently defaulting to "covered".
	return lacks_coverage is None or bool(lacks_coverage)


async def evaluate_route_hazards(
	db: AsyncSession,
	waypoints: Sequence[Waypoint],
	safety_margin_m: float,
	safety_contour_m: float,
) -> ValidationResult:
	route_wkt = _route_linestring_wkt(waypoints)
	if route_wkt is None:
		return ValidationResult(status="safe", hazards=[])

	blocked_hazards: list[HazardHit] = []
	for layer, description in _BLOCKED_LAYER_DESCRIPTIONS.items():
		count = await _count_nearby(db, f"enc_{layer}", route_wkt, safety_margin_m)
		if count > 0:
			blocked_hazards.append(HazardHit(layer=layer, description=description, count=count))

	if blocked_hazards:
		return ValidationResult(status="blocked", hazards=blocked_hazards)

	# Below "blocked": collect every applicable hazard (coverage gap, depth warning) rather than
	# stopping at the first, so the operator sees the full picture -- a route can be no_data along
	# one leg and shallow along another. Overall status is the worst tier present: no_data outranks
	# warning (an unknown risk deserves at least as much attention as a known, tolerable one), but
	# neither blocks send_mission() -- only "blocked" does.
	hazards: list[HazardHit] = []
	lacks_coverage = await _route_lacks_coverage(db, route_wkt, safety_margin_m)
	if lacks_coverage:
		hazards.append(HazardHit(layer="m_covr", description=_NO_DATA_DESCRIPTION, count=1))

	depth_count = await _count_nearby(
		db,
		f"enc_{_DEPTH_LAYER}",
		route_wkt,
		safety_margin_m,
		max_depth_m=safety_contour_m,
	)
	if depth_count > 0:
		hazards.append(
			HazardHit(
				layer=_DEPTH_LAYER,
				description=(
					f"Route crosses charted depth below the {safety_contour_m} m safety contour."
				),
				count=depth_count,
			)
		)

	if lacks_coverage:
		return ValidationResult(status="no_data", hazards=hazards)
	if depth_count > 0:
		return ValidationResult(status="warning", hazards=hazards)
	return ValidationResult(status="safe", hazards=[])
